### PR TITLE
Time independent skulltulas

### LINKED ISSUE #2356: Time Dependant Skulls Always Appear
[Migrated from Trello](https://trello.com/c/B4QTV1H2)

Setting that makes Gold Skulltulas that normally appear only at night available any time of day.

### PR DESCRIPTION
**Pull request by mracsys**
Implements #2356.

Two parts to this PR. There is the skulltula time of day feature, and a new framework to enable arbitrary changes to any of the scene or room files parsed to linked python objects. The framework automatically shifts all pointers within the files to handle new file locations and additions/deletions within the file. It will have multiple applications besides the skulltula feature:

* Detailed cutscene parsing instead of writing raw bytes (WIP). Also more easily allows writing completely new cutscenes.
* Adding any actor/object to any room, such as a sign outside Ganon’s castle for the bridge condition or anywhere else a hint might be helpful
* Entrance/exit additions, such as promoting grottos to the same framework as other entrances, or new loading planes for mixed pool door rando
* Space reclamation. Almost a MB is wasted in padding between scene and room files.
* More user friendly interface for MQ patching, and opens the door for other alternate scene options for anyone crazy enough to attempt it.
* Collision changes for problems like #2331 and for my vision of door rando without scene reloads between doors.

More details on the design of the parser is included in `Notes/scene-and-room-patching.md`.

# GUI Impacts for Web

TL;DR, `Creating Patch File` is added as a progress bar milestone between `Patching ROM` and `Starting compression`.

Due to the increased time spent patching the rom caused by the significantly larger amount of changed bytes compared to the current randomizer, an additional status message milestone is added to the GUI progress bar: `Creating Patch File`. This was previously included in the `Patching ROM` milestone. Additionally, as these milestones alternate for each player in multiworld generation, the progress percentages for these two milestones were modified from a static number to one that scales based on the currently processing world. This requires resetting `currentWorld` after successful patching for subsequent generations to work when `Generation Count` is greater than 1.

# Python Virtual Environment Addition

This PR makes use of the [numpy](https://pypi.org/project/numpy/) package to more efficiently process changed bytes for patch generation compared to base python lists and dictionaries. While the randomizer has avoided external dependencies thus far, using numpy reduces the performance penalty substantially (times are from single world runs on an AMD 5900X with the S8 Tournament preset):

| Task | Penalty with numpy | Penalty without numpy |
| :----: | :----: | :----: |
| ROM Generation | 25% (1.6s) | 66% (4.3s) |
| ZPF Generation | 168% (8.4s) | 278% (13.9s) |

A virtual environment is automatically created and activated by the major entry points to the randomizer:

* **GUI** - `GUI.py`
* **Console** - `OoTRandomizer.py`
* **CI** - `CI.py`

Each of these entry points independently check if they are running in a virtual environment located in subfolder `.venv` at the local path. If the environment doesn't exist, it is created automatically. There is also check if all dependencies in `requirements.txt` are fulfilled via pip. If not fulfilled, they are installed automatically in the virtual environment. If running in a release build, any pip errors are captured and passed to the regular error popup. Otherwise, running via the scripts logs errors to the console before exiting. If venv and pip checks are successful, but either the current python environment is not the virtual environment, or new dependencies were added, the script relaunches itself via the `subprocess` module in the virtual environment with all arguments intact. This behavior can be added to additional entry points or custom scripts by importing the `ensure_venv()` function from `Utils.py` and adding it as the first line wherever it's needed. See `gui_main()`, `start()`, and the top level of `CI.py` for example usage.

numpy is pinned between the last version supported by python 3.8 (1.24.4) and the version in use on RealRob's `custom_voice_packs` branch (2.2.1). Both versions have been tested to work. For compatibility with RealRob's work, I strongly recommend increasing the minimum python version for the randomizer to 3.10, the earliest supported by numpy 2.2.1. 3.8 is EOL, and 3.9 will be EOL in October 2025. This PR does not change the minimum python version.

# Unrelated Bugfixes

* Fix a typo in an exception for wad file generation (see https://github.com/OoTRandomizer/OoT-Randomizer/blob/a8217053ce68884c9950ab91cf4bb66029a964b2/Main.py#L267)
* Fix `Rom.copy()` handling of the `original` property. On copy, the new object's `original` would refer to itself instead of the unmodified rom buffer. This was causing problems re-using the patched rom before cosmetics to output a rom and zpf at the same time. This isn't a problem with the current randomizer as it always fully repatches the rom for every output. The same issue also applied to the `overlay_table`, but that had no ill effects.

# Todo

- [X] Test compressed roms. ~~DMA changes for compressor?~~
- [X] Update pointers outside scene files to globals in them, mainly cutscenes.
- [X] If 2 doesn’t fix problems moving file start addresses, figure that out
- [X] Integrate framework with rando patching pipeline
- [X] Move cutscene patches to framework
- [X] Move MQ patching to framework
- [X] Move any other scene changes to framework
- [X] Add setting for skulltulas ignoring time of day. Add actors/objects where needed and patch all actor parameters.

# Rando Patches to Test with New System

- [X] Cutscene Patching
    - [X] Cutscenes
      - [X] Lost Woods Bridge Saria's Gift
      - [X] Song cutscenes without songs as items
      - [X] Song cutscenes with songs as items
      - [X] Open Royal Family Tomb as both adult/child
      - [X] Darunia's Dance for Saria's Song
      - [X] Owl warps
      - [X] Zelda escaping from Hyrule Castle for OoT check
      - [X] Small cutscene after learning OoT song
      - [X] Epona race start
      - [X] Epona escapes to different Hyrule Field entrances
      - [X] Burning Kak intro cutscene
      - [X] Well draining cutscene (Kakariko)
      - [X] Nabooru knuckle defeat
      - [X] Rainbow bridge
      - [X] Trial completion cutscenes
      - [X] Ganon's Tower collapse
      - [X] Phantom Ganon blue warp Deku Sprout cutscene skip
    - [X] Cutscenes outside scene files
      - [X] Jabu Jabu swallowing Link
      - [X] Ruto pointing to dungeon reward in Big Octo room
      - [X] Opening Door of Time
      - [X] Master Sword pedestal cutscene
      - [X] Well draining cutscene (Windmill)
    - [X] wondertalk2 actor moves
      - [X] Shadow Temple whispering maze (8x)
      - [X] Shadow Temple Truthspinner (2x)
      - [X] GTG Entrance (3x)
      - [X] GTG Stalfos room (1x)
      - [X] GTG Flame Wall Maze/Slopes Room (1x)
      - [X] GTG Pushblock Room (2x)
      - [X] GTG Rotating Statue Room (1x)
      - [X] GTG Megaton Statue/Back Enemies Room (1x)
      - [X] GTG Lava Room (3x)
      - [X] GTG Dinolfos Room (1x)
      - [X] GTG Inner Maze (1x)
      - [X] GTG Shellblade Room/Toilet (1x)
      - [X] Death Mountain Crater (1x)
      - [X] Hideout Cells (1 per cell room)
- [x] Other Patches
    - [X] Duplicate Bazaar room for Kakariko
    - [X] Move Sheik from pedastal in ToT
    - [X] Ice Cavern alcove camera (duplicated in two spots in Patches.py)
    - [X] Fire Temple boss loop unlocked door without keysanity
    - [X] Non-MQ Water Temple door always unlocked
    - [X] Graveyard ledge grabs
    - [X] Owl removals
    - [X] Jabu octorok position
    - [X] Forest/Fire Temple switch heights
    - [X] Kakariko carpenter starting position
    - [X] Vanilla DC gossip stone fairy flag
    - [X] Colossus Fairy entrance "...???" text
    - [X] Forbid Sun's song in a bunch of cutscenes
    - [X] Move Fado for adult trade shuffle
    - [X] Spirit Shortcut actor tweaks
    - [X] Gerudo Fortress gate guard reposition
    - [X] Skip child stealth crawlspace exit
    - [x] Silver rupee shuffle MQ DC/Spirit temp->permanent flags
    - [X] ~~Well ladder rupee reposition~~ No longer relevant as of #2551
    - [X] Shadow Temple redead shared flags for silver rupee shuffle
    - [X] Song shuffle cutscene text boxes
    - [X] Song shuffle location addresses
    - [X] Shopsanity shop item objects in room headers
    - [X] Cow shuffle repositions
        - [X] Stable
        - [X] Tower
        - [X] Shuffled item actor params in set_cow_id_data
    - [X] CSMC repositions
        - [X] Vanilla Ganons Castle Light Trial
        - [X] Vanilla Spirit Temple compass chest
        - [X] Silver Gauntlet chest in glitched logic
    - [X] Dead Hand spawns in vanilla Shadow/Well
    - [X] Broken drops vanilla Spirit deku shield in anubis room
    - [X] TCG shuffle temp flags to permanent/keysy
    - [X] Remove "entrance blockers"
    - [X] Scrub shuffle actor params in set_deku_salesman_data
    - [X] Jabu stone actor
    - [X] Keysy dungeon/boss doors
    - [X] Ganons Tower first BK door unlock for pot shuffle
- [x] Entrance Shuffle
    - [X] Generate/write exit list for each scene
    - [x] Jabu boss exit coordinates
    - [x] Water Temple boss exit room number
    - [X] Redirect LLR exits to main exit for OW shuffle
    - [X] Redirect ZR<->Field exits to land from water
    - [X] Spirit temp flag purge on entry through front door
    - [X] Grotto actor data changes in set_grotto_shuffle_data
- [X] MQ Patching
    - [X] All rooms can be entered from all entrances
      - [X] Deku Tree
      - [X] Dodongo's Cavern
      - [X] Jabu Jabu's Belly
      - [X] Forest Temple
      - [X] Fire Temple
      - [X] Water Temple
      - [x] Spirit Temple
      - [X] Shadow Temple
      - [X] Bottom of the Well
      - [X] Ice Cavern
      - [X] Gerudo Training Ground
      - [x] Ganon's Castle
    - [X] Ice Cavern scene header patch from old system
    - [x] MQ Spirit Temple room 6 new alternate header
    - [X] Shadow Temple MQ redead shared flags for silver rupee shuffle
    - [x] DC MQ door flag move for silver rupee shuffle
    - [x] Spirit Temple MQ front right chest temp -> permanent flag for silver rupee shuffle
    - [X] Key doors are correct
    - [X] Check overrides work
    - [X] Scrub actor patching works
    - [X] Cow actor patching works
    - [X] Jabu MQ dungeon reward model override works
    - [X] Keysy removes locks
- [X] ASM/C patches
    - [X] 0x26c10e3 - generic grotto ACTOR_EN_GS actor params 0x3818 -> 0x38FF (use grotto ID for hint text ID)

### CODE REVIEW

**Comment by r0bd0g** [comment]
I guess for logic you'd go around to each skulltula you affected that checks for night and change at_night to (at_night or [setting]). Something to look out for is the watchtower, which will no longer be climbable in child day without a means of killing the skulltula. In closed forest, the time of day access check is skipped, so there's special logic on the night skull there to confirm you can actually get it to night somehow, and you'd also have to add the new setting as an alternative for that.

File: ProcessActors.py [hunk 1]
```diff
@@ -360,15 +583,125 @@ def process_pot(actor_bytes):
...
+
+                        if init_obj not in room.setups[setup].objects and init_obj != scene.keep_id and init_obj != 1 and init_obj <= 0x0191: # Check if the init object is in the rooms data
+                            bad_actors.append((actor.addr, scenes[scene.id], scene.id, room.id, setup, i, actor.id, overlay_entry['init_object']))
+                        i += 1
+    for actor in bad_actors:
+        print(actor)
+    return bad_actors
+
+
+#rom = Rom("ZOOTDEC.z64")
```
Comment (mzxrules) [inline ProcessActors.py]: what's with all the dead code here

File: Scene.py [hunk 2]
```diff
@@ -0,0 +1,2539 @@
...
+    while i < original_size:
+        if original_bytes[i] != new_bytes[i]:
+            raise Exception(f'Byte mismatch at offset 0x{i:0>8x}. Original: 0x{original_bytes[i]:0>2x} New: 0x{new_bytes[i]:0>2x}')
+        i += 1
+
+
+if __name__ == '__main__':
+    uncompressed_rom = Rom('ZOOTDEC.z64')
+    compare_parsed_data_to_rom(uncompressed_rom, True)
+    #check_external_reference_locations(uncompressed_rom)
```
Comment (mzxrules) [inline Scene.py]: more dead code

File: Cutscenes.py [hunk 3]
```diff
@@ -485,33 +634,1313 @@ def patch_cutscenes(rom: Rom, songs_as_items: bool, settings: Settings) -> None:
...
-        rom.write_int16s(0x2059412, [0x03C0, 0x00E2, 0xFAA6]) # Move coordinates. Original value : 1000, 205, -1202. New value : 960, 226, -1370.
-        rom.write_byte(0x205941F, 0xBF) # Never set the flag.
+        actor = scenes[SceneIDs.ZORAS_RIVER].rooms[0].headers[0].actor_list.actors[58]
+        actor.pos = Vec3s(960, 226, -1370) # original 1000, -205, -1202
+        actor.params = 0x4BBF # original 0x4BBB
+
+
+# Gaps in IDs are intentional
+# https://github.com/zeldaret/oot/blob/7235af2249843fb68740111b70089bad827a4730/include/z64cutscene.h#L35-L165
+class CutsceneCommandID(IntEnum):
```
Comment (mzxrules) [inline Cutscenes.py:710]: man, rando sure is a hot mess when it comes to organization. 

I find it kind of odd to just stuff these important type definitions after a bunch of function patching routines split out of patches.py. I think we need to consider moving this type access stuff into a new folder like "ztypes" or something

Comment (mracsys) [reply Cutscenes.py:710]: You're right, I should have put more effort into structure. Some of the other utility classes like `Vec3s` and `Vec3i` that didn't fit well got put into Rom.py despite not fitting there either. I can refactor into the same file/folder structure as decomp for structs lifted from there, with a top level folder to keep it neat.

File: ProcessActors.py (hunk 1 above)
Comment (mracsys) [reply ProcessActors.py]: This is a utility script from @rrealmuto used on his branch. Some of this was already in the main repo. I can remove the whole file if needed and its not inconvenient to Rob.

File: Scene.py (hunk 2 above)
Comment (mracsys) [reply Scene.py]: removed

**Comment by r0bd0g** [comment]
I warned you about special logic that would be needed for KF GS Know It All House! As you have it currently, in closed forest, if you start at daytime, under this new setting the skulltula will be there but you won't have logic to collect it before beating deku or finding sun's song. That skull has an extra check on it to make sure it's actually present since closed forest breaks the assumption that you will always have access to moving time of day from spawn, so you need to add the new setting as an alternative way to have the time of day to access it.

(This isn't really related to your PR, you shouldn't change anything about this, but I wonder if all of those skulltulas that check for ToD in areas where time moves already, should just have the time of day check removed. It was nice to look at before but with the new setting it's become kind of bloated. I made a choice along those lines with Peahats when reviewing enemy drop shuffle b/c separate night and day logic for them was pretty convoluted for no real benefit.)

I don't know enough about glitched to say whether or not the logic is wrong absolutely for sure or not, but at a glance I don't trust it. For example, man on roof, you check for a weapon to kill the skulltula, but you always are checking it regardless of if the skulltula is potentially there at daytime. You should get a glitched logic expert to review it, or make no glitched logic changes and mark the settings as incompatible.

(Unrelated to your PR but I was glancing over the night skulls in glitched logic and noticed some logic for buying the Giant's Knife that is assuming that the Knife isn't shuffled. This shows the dangers of marking new settings as glitched logic compatible without getting a specialist's review!)

**Comment by fenhl** [comment]
Should this setting be enabled in the Easy Mode preset?

**Comment by mracsys** [comment]
It would make sense. Would that need to go through some sort of vote or poll?

**Comment by fenhl** [comment]
Current dev team call is in favor so I'd say you can just add it.

**Comment by shirosoluna** [comment]
@r0bd0g > 
> I don't know enough about glitched to say whether or not the logic is wrong absolutely for sure or not, but at a glance I don't trust it. For example, man on roof, you check for a weapon to kill the skulltula, but you always are checking it regardless of if the skulltula is potentially there at daytime. You should get a glitched logic expert to review it, or make no glitched logic changes and mark the settings as incompatible.
> 
> (Unrelated to your PR but I was glancing over the night skulls in glitched logic and noticed some logic for buying the Giant's Knife that is assuming that the Knife isn't shuffled. This shows the dangers of marking new settings as glitched logic compatible without getting a specialist's review!)

I wouldnt worry too much about glitched logic right now, because it's being updated entirely to have all settings included. There are way more important stuff that isnt compatible at all currently and in released versions. If this is going to be slated for 9.0 as well then we can make it incorporated. I dont know the timing for this PR. 
Just offering support to not be concerned about glitched logic too too much atm

**Comment by fenhl** [comment]
There's not really any timing for this PR, it will be merged once it's been code reviewed and sufficiently tested. Since it's a rather large change, that might take a while, but so is glitched logic 3.0, so it's hard to say which is more likely to be merged first. I'd recommend not including the changes for this in glitched logic 3.0 for now or perhaps maintaining them separately.

**Comment by shirosoluna** [comment]
> There's not really any timing for this PR, it will be merged once it's been code reviewed and sufficiently tested. Since it's a rather large change, that might take a while, but so is glitched logic 3.0, so it's hard to say which is more likely to be merged first. I'd recommend not including the changes for this in glitched logic 3.0 for now or perhaps maintaining them separately.

wasn't meaning that it would wait for it but that I'd incorporate it into logic when the time comes. Even if it's after being pulled. 
They are maintained separately. But if new glitch logic comes later then still dont bother worry about current glitched logic. It's too incompatible with everything currently and one more setting wouldnt make a difference lol.

File: ASM/src/hacks.asm [hunk 4]
```diff
@@ -2584,8 +2584,8 @@ skip_bombchu_bowling_prize_switch:
 .orga 0xEE7E4C
     jal Hit_Gossip_Stone
 
-.orga 0x26C10E3
-    .byte 0xFF ; Set generic grotto text ID to load from grotto ID
+;.orga 0x26C10E3
+;    .byte 0xFF ; Set generic grotto text ID to load from grotto ID
```
Comment (fenhl) [inline ASM/src/hacks.asm]: Any reason this is commented out? If it's not necessary anymore, it'd be better to just remove it.

File: Patches.py [hunk 5]
```diff
@@ -1441,32 +1462,38 @@ def calculate_traded_flags(world):
...
-        rupeetower_locations = [location for location in world.get_locations() if location.disabled == DisableType.ENABLED and location.type == 'RupeeTower']
+    # if world.settings.shuffle_freestanding_items:
+    #     # Get freestanding item locations
+    #     actor_override_locations = [location for location in world.get_locations() if location.disabled == DisableType.ENABLED and location.type == 'ActorOverride']
+    #     rupeetower_locations = [location for location in world.get_locations() if location.disabled == DisableType.ENABLED and location.type == 'RupeeTower']
 
-        for location in actor_override_locations:
-            patch_actor_override(location, rom)
+    #     for location in actor_override_locations:
+    #         patch_actor_override(location, rom)
```
Comment (fenhl) [inline Patches.py]: This can be removed, it's no longer needed since these were refactored in potsanity 3.0.

File: ProcessActors.py (hunk 1 above)
Comment (fenhl) [reply ProcessActors.py]: Rob says it's fine to remove it.

**Review by fenhl — CHANGES_REQUESTED** [review]

**Comment by fenhl** [comment]
I'll see if we can get a web generator for this branch set up to encourage more testing.

File: ASM/src/hacks.asm (hunk 4 above)
Comment (mracsys) [reply ASM/src/hacks.asm]: Nope, missed this when I was cleaning up all the old stuff I kept for reference.

**Comment by mracsys** [comment]
FYI, I don't have much free time through at least fall, possibly into next year. I'm still around, but changes might be delayed.

**Review by fenhl — APPROVED** [review]

**Comment by fenhl** [comment]
This PR branch is now available as a web generator at <https://ootrandomizer.com/generatorDev?version=devM_> (ignore the “currently unmaintained” in the title, it is up to date with the PR). Any feedback from people testing it would be appreciated, even if that feedback is “everything worked well”.

**Comment by mracsys** [comment]
I'm done with changes to this pending additional reviews.

File: Gui.py
Comment (rrealmuto) [inline Gui.py:1]: Came here because I wanted to steal your virtual environment code. I think I had mentioned this previously but checking for/creating the venv in Gui.py is not sufficient. Gui.py is not invoked in the release build. Instead electron is started directly using node.

I have a semi-working implementation of creating the venv and updating packages from javascript in #2143 which generally works with the release build. However the lack of solid error handling has been what has stopped me from merging this on my branch already

Comment (fenhl) [reply Gui.py:1]: Do you think you two could combine your efforts to add something that'll work for both use-cases to this PR?

**Review by fenhl — CHANGES_REQUESTED** [review]
@rrealmuto's concerns about the venv handling still need to be addressed. Other than that, this looks good to me.

**Comment by mracsys** [comment]
@rrealmuto Can you point me to where the problem is? I built the installer in a fresh Windows environment and had no problems with the installed release build making it's own venv in its appdata folder. Also tested on Linux. requirements.txt needs to be manually copied into the GUI/python subfolder on build along with all the other files.

The function to create the venv lives in Gui.py, but it's called in OoTRandomizer.py and CI.py as well to handle those use cases.

**Comment by mracsys** [comment]
Error reporting if pip fails is handled for dev and release. An error is thrown with the pip stderr output from python, which is captured by the GUI when starting the application (independent of launching through `Gui.py`). I tested it by disconnecting from the internet and removing the `.venv` folder and attempting to launch. The GUI has a loading icon preventing interaction on everything except the patch area at the bottom until the venv is either created or fails (with error display). I can't test failures on web, though if I understand the architecture there correctly, the venv is created once on first visit for all users. Any errors there would be quickly apparent on branch update, and presumably would have the immediate attention of the dev pushing the branch.

@rrealmuto does this address your concerns?

I also rebased to 9.x, which includes integrating advanced logic. @shirosoluna would you mind taking a look to make sure I got everything right?

**Comment by fenhl** [comment]
I pinged @rrealmuto on Discord and this was the reply:

> I read his comment. Still don't think any of the venv code should be done in Gui.py this is literally exactly what trez and DB told me not to do.

> Gui.py doesn't get called by the release build

> I have a (somewhat) robust venv setup that gets handled in the electron code. Displays status during the setup and reports errors on the GUI. 
>
> Both of these solutions run into the issue that you won't have the dependencies if you don't launch the GUI. So they don't work for 1) the web and 2) any CLI users without first having the dependencies installed

Personally I don't think that's a big issue since we can add instructions for CLI users to install the deps, and we're currently discussing how to make things work on the website in a disk-space-efficient manner. (Backwards-incompatible dependency upgrades will likely need to be kept to a minimum so venvs can be reused across randomizer versions.)

**Comment by mracsys** [comment]
Like I've said a few times, the venv setup is called by the CLI entry scripts (Ootrandomizer.py and CI.py). Gui.py also runs it independent of any of the CLI entries for anyone using the randomizer that way, but it is not the only path for venv creation. I've tested all of this on fresh python installs for CLI, GUI, and release builds. None of it depends solely on the GUI. The function is imported from Gui.py because it has to live somewhere. I can move that to Ootrandomizer.py, Rom.py, Main.py, or wherever is preferred.

I can add a version header to the requirements.txt file in a comment if that makes web deduplication easier.

**Comment by flagrama** [comment]
I think the most appropriate existing file would be Utils. Otherwise a new file.

I definitely do not expect to find virtual env information in Gui.py or Ootrandomizer.py. Including Main.py just sounds like bad practice (regardless of if we already do it elsewhere, I did not check).

**Comment by rrealmuto** [comment]
Talked to mracsys and no more concerns on the venv bit

**Comment by mracsys** [comment]
I need to apologize for getting snippy in my last message. I wasn't understanding what I was failing to communicate, but that's not an excuse. Things resolved quickly on Discord. I expanded more on the venv system in the PR description for posterity. Thanks again Rob!

**Review by fenhl — APPROVED** [review]

**Comment by fenhl** [comment]
Thank you for sorting this out. I should apologize as well for not realizing what the misunderstanding was, I could have caught that if I'd paid closer attention.

The website is being updated with venv support and I've requested an update for the devM generator so I can call for another round of testing.

File: data/Glitched World/Overworld.json [hunk 6]
```diff
@@ -1989,7 +1989,7 @@
                 or can_hover
                 or (can_use(Hover_Boots) and (can_mega or can_hess or can_superslide))",
             "Kak Odd Medicine Rooftop": "is_adult
-                or (is_child and (at_day or
+                or (is_child and ((at_day and not gold_skulls_ignore_daytime) or
```
Comment (shirosoluna) [inline data/Glitched World/Overworld.json:1994]: I think this was added in error? I think you may have just looked for at_day and added the skull helper. This one can be removed since it is just access to the area not skull related

Comment (fenhl) [reply data/Glitched World/Overworld.json:1994]: This is for getting onto granny's potion shop's roof via the watchtower, so the skull is in the way.

Comment (shirosoluna) [reply data/Glitched World/Overworld.json:1994]: otherwise no comments from me! thank you for adding advanced logic ❤️

Comment (shirosoluna) [reply data/Glitched World/Overworld.json:1994]: > This is for getting onto granny's potion shop's roof via the watchtower, so the skull is in the way.

ah I see. I was scanning it for the is_adult and is_child making sure they're differentiated. And didnt notice the "not" part. Thanks for clarifying

**Comment by fenhl** [comment]
Since there were no responses to our previous call for testing, we're going to try posting another one in the new #dev-announcements channel for more visibility. Before we do, could you rebase the PR (or merge Dev) please?

File: Cutscenes.py [hunk 7]
```diff
@@ -147,109 +151,165 @@ def patch_cutscenes(rom: Rom, songs_as_items: bool, settings: Settings) -> None:
...
+    delete_cutscene(darunia_cutscene)
 
     # Speed up Death Mountain Trail Owl Flight
-    patch_cutscene_destination_and_length(rom, 0x223B6B0, 1)
+    dmt_owl_cutscene = scenes[SceneIDs.DEATH_MOUNTAIN_TRAIL].get_existing_record_by_vanilla_offset(0x1E6A0, RecordType.CutsceneData)
+    patch_cutscene_destination_and_length(dmt_owl_cutscene, 422, 1)
 
     # Jabu Jabu swallowing Link
-    patch_cutscene_destination_and_length(rom, 0xCA0784, 1)
+    jabu_swallow_cutscene = Cutscene.decode(rom, 0xC9FC84)
```
Comment (cjohnson57) [inline Cutscenes.py:192]: Not something that needs to be fixed really but it would be nice if these "magic numbers" could be stored in a table too, like how many of the others have been eliminated by these changes

File: Rom.py [hunk 8]
```diff
@@ -140,25 +157,50 @@ def decompress_rom(self, input_file: str, output_file: str, verify_crc: bool = T
...
+
+    def read_s16(self, address: Optional[int] = None) -> int:
+        if address is None:
+            address = self.last_address
+        return int.from_bytes(self.read_bytes(address, 2), 'big', signed=True)
+
+    def read_s32(self, address: Optional[int] = None) -> int:
+        if address is None:
+            address = self.last_address
+        return int.from_bytes(self.read_bytes(address, 4), 'big', signed=True)
```
Comment (cjohnson57) [inline Rom.py:173]: Is there a particular reason not to make these one function with the size as a param? (Though that being said, I think I might have done the same thing in my model code)

**Review by cjohnson57 — APPROVED** [review]
No real blockers, just a couple small things I noticed. Though, I feel a little impotent in my ability to review this properly, given the scope and the fact that there's thousands of lines of transliterated game data that I can't really have an opinion on.

I kind of wonder if it's worth removing the nighttime skulls expect sun song option now... It feels a bit vestigial. Doesn't do the same thing obviously, but fulfills the role of "I don't want to mess with waiting for night for skulls)

Also, I see this is still tagged as needs testing, is this actually still the case?

File: Rom.py (hunk 8 above)
Comment (flagrama) [reply Rom.py:173]: This is pretty standard when working with binary data. Having to specify the size as a parameter just makes each call harder to read because you have to think about which one is the size parameter and may have to refer to the function definition, but by having separate calls for each size makes it obvious. Especially with these being for signed bytes I think it is very important to make it absolutely obvious what the function is doing.

**Comment by fenhl** [comment]
> Also, I see this is still tagged as needs testing, is this actually still the case?

No, I consider this to be sufficiently tested after the recent full-game test. I guess I forgot to remove the label.

File: Cutscenes.py (hunk 7 above)
Comment (mracsys) [reply Cutscenes.py:192]: Agreed it could be better documented. Cutscenes are stored all over the place in the codebase with no definitive centralized list unfortunately. This one is in the Jabu actor file for example and only called by it.

The cutscene objects try to minimize the magic VROM locations to just the start of the cutscene. Those are easy to line up with the decomp ntsc-1.0 map file for the cutscene variable address. In the future, maybe there could be a writeup at the top of the file on how to do that.

File: Rom.py (hunk 8 above)
Comment (mracsys) [reply Rom.py:173]: Basically what flag said. With how often these are called in the parsing functions, it's more ergonomic to do `rom.read_s16(address)` than something like `rom.read_bytes_to_int(address, length, signed=True)`. Also makes it more obvious in the calling code what the C type of the data is supposed to be.

**Comment by mracsys** [comment]
> > Also, I see this is still tagged as needs testing, is this actually still the case?
> 
> No, I consider this to be sufficiently tested after the recent full-game test. I guess I forgot to remove the label.

To be clear, I've had multiple full-game tests of this on console and emulator for a while in addition to all the individual cases checked off in the PR description. Always preferable to have independent testing though.

**Comment by fenhl** [comment]
…I was not aware of that.

**Comment by mracsys** [comment]
I'm not sure I ever said that to be fair, looking back at the description.

**Comment by fenhl** [comment]
Oh well, like you said, doesn't hurt to have independent confirmation too.

**Comment by fenhl** [comment]
I'm having trouble merging this because my pre-commit script tries to run unit tests which fail with errors like:

```
======================================================================
ERROR: test_audiobank (Unittest.TestCustomAudio.test_audiobank)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/september-shared/git-stages/github.com/OoTRandomizer/OoT-Randomizer/Unittest.py", line 1052, in test_audiobank
    rom: Rom = Rom("ZOOTDEC.z64")
               ~~~^^^^^^^^^^^^^^^
  File "/mnt/september-shared/git-stages/github.com/OoTRandomizer/OoT-Randomizer/Rom.py", line 46, in __init__
    self.changed_address: numpy.ndarray = numpy.full(134217728, 1000, numpy.uint16)
                                          ^^^^^
NameError: name 'numpy' is not defined
```

This is presumably a quirk of how Python works on NixOS, as adding numpy to the Python packages in the Nix devShell (which I intend to PR in the future) and skipping the venv works:

```nix
python3.withPackages (python-pkgs: with python-pkgs; [
    numpy
])
```

Any objections to adding a `--no-venv` CLI flag that makes `Utils.ensure_venv` return immediately?

**Comment by mracsys** [comment]
Nope, go for it!
